index.bs

```diff
@@ -5506,6 +5729,329 @@ partial interface MLGraphBuilder {
+
+### Creating {{MLOperand}} ### {#api-mloperand-create}
+The {{MLOperand}} objects are created by the methods of {{MLGraphBuilder}}, internally using the following algorithms.
+
+<details open algorithm>
+  <summary>
+    To <dfn>create an MLOperand</dfn> given |builder| and |desc|, run the following steps:
+  </summary>
+  <div class=algorithm-steps>
+    1. [=Assert=]: the type of |builder| is {{MLGraphBuilder}}.
+    1. [=Assert=]: the type of |desc| is {{MLOperandDescriptor}}.
+    1. Let |operand| be a new [=object=].
+    1. Set |operand|.{{MLOperand/[[builder]]}} to |builder|.
+    1. Set |operand|.{{MLOperand/[[descriptor]]}} to |desc|.
+    1. Return |operand|.
+  </div>
+</details>
+
+<details open algorithm>
+  <summary>
+    To <dfn>copy an MLOperand</dfn> given |operand|, run the following steps:
+  </summary>
+  <div class=algorithm-steps>
+    1. [=Assert=]: the type of |operand| is {{MLOperand}}.
+    1. Let |result| be a new [=object=].
+    1. Set |result|.{{MLOperand/[[builder]]}} to |operand|.{{MLOperand/[[builder]]}}.
+    1. Set |result|.{{MLOperand/[[descriptor]]}} to |operand|.{{MLOperand/[[descriptor]]}}.
+    1. If |operand|.{{MLOperand/[[name]]}} [=map/exists=], then set |result|.{{MLOperand/[[name]]}} to |operand|.{{MLOperand/[[name]]}}.
+    1. Return |result|.
+  </div>
+</details>
+
+<details open algorithm>
+  <summary>
+    To <dfn>check dimensions</dfn> given |dimensions| and |type|, run the following steps:
+  </summary>
+  <div class=algorithm-steps>
+    1. If the [=list/size=] of |dimensions| is 0, return false.
+    1. If the [=list/size=] of |dimensions| is too large to be supported by the implementation, return false.
+    1. If any element of |dimensions| is not a positive number, or it is too large to be supported by the implementation given |type|, return false.
+    1. Return true.
+  </div>
+</details>
+
+<details open algorithm>
+  <summary>
+    To <dfn for="MLOperand">validate MLOperand</dfn> given |operand| and |builder|, run the following steps:
+  </summary>
+  <div class=algorithm-steps>
+    1. [=Assert=]: the type of |operand|.{{MLOperand/[[builder]]}} is {{MLGraphBuilder}}.
+    1. If |builder| is not equal to |operand|.{{MLOperand/[[builder]]}}, return false.
+    1. Let |desc| be |operand|.{{MLOperand/[[descriptor]]}}.
+    1. If |desc|.{{MLOperandDescriptor/dimensions}} [=map/exists=] and invoking <a>check dimensions</a> given |desc|.{{MLOperandDescriptor/dimensions}} and |desc|.{{MLOperandDescriptor/type}} returns false, then return false.
+    1. Return true.
+  </div>
+</details>
+
+## {{MLOperandDescriptor}} dictionary ## {#api-mloperanddescriptor}
+<script type=idl>
+enum MLInputOperandLayout {
+  "nchw",
+  "nhwc"
+};
+
+enum MLOperandDataType {
+  "float32",
+  "float16",
+  "int32",
+  "uint32",
+  "int64",
+  "uint64",
+  "int8",
+  "uint8"
```

Comment:
Should the data types be listed in an order like ascending (..., int8, uint8, int32, uint32, int64, uint64) or descending (..., int64, uint64, int32, uint32, int8, uint8) or ascending/descending grouped (..., int8, int32, int64, uint8, uint32, uint64)? As discussed in [this CL review](https://chromium-review.googlesource.com/c/chromium/src/+/5031526/3/components/ml/webnn/graph_validation_utils.h#34).

Reply:
The sorted enum list in the CL is a bit different though since it's C++ and that the position of a C++ enum value reflects the integer indices of the enum values. The WebIDL enum values are string-based whose relative positioning i.e. the orde in which they are defined doesn't matter. I agree sorted names makes the list looks nicer overall but it also means we can't have "float32" as the first entry since it's sorted after "float16", and if we have "float8" in the future, the first entry will get reshuffled again. I don't feel strongly that an action is needed here.

Reply:
With confirmation (👍) from @shiyi9801, I will mark this conversation resolved.